### DISCUSSION TITLE
How to get all required attributes from product type definitions JSON schema

### DISCUSSION THREAD

**Discussion by serdargokhan** [amzn/selling-partner-api-models#3713] [General] [answered]
I am integrating an external system to publish products on Amazon. I get type definitions for specific categories from the product type definitions API, which gives me the schema.

The problem is that I want to get all the required attributes and enums of those attributes ahead of time. But the schema is so complex. What is the solution to this specific problem? I appreciate your interest!

**Comment by vaidas-lungis** [comment, marked as answer]
Complex it is  😅 
Especially if you are creating new listings. IMHO its iterative process build a payload and send it to the API to get initial validation. OFC you can do validation yourself locally before sending to Amazon Listing API with some JSON Schema validation tools (or packages depending on your programming language). So you will cut some additional HTTP roundtrips and lower the request to the API. Hence local validation results could give some different results than the Amazon Listing API response ;) 

You can always check the Amazon's recommended workflow https://developer-docs.amazon.com/sp-api/docs/building-listings-management-workflows-guide 

Allso keep in mind of editable attributes later after new listing successfully created. 
https://github.com/amzn/selling-partner-api-docs/discussions/1609

**Reply by serdargokhan** [reply]
Okay, it seems there are no easier ways :( Thank you for your help.

**Reply by mahmoudAhme-d** [reply]
Did you solve the problem ?

**Reply by serdargokhan** [reply]
Nah, we gave up and decided not to upload products but only update stocks and pricing.

**Reply by weilinggu** [reply]
Hi @serdargokhan Adding some information that might be helpful. The required attributes on a product type are dynamic. As you fill out different attributes, the required attributes might change. You can find `allOf` in product type JSON schema returned, which contains the conditional requirements.

A few resources that might help:
- [Here](https://youtu.be/fraQT3UhFII?si=K4tCPUOA3gILc7mU) is a deep dive video on product type JSON schema that explains how the product type JSON schema works with a demo submission in part II
- [Here](https://github.com/amzn/selling-partner-api-samples/tree/main/use-cases/listings) is a Listings sample application that implemented a dynamic UI based on product type JSON schema.

[json-schema.org](https://json-schema.org/tools?query=&sortBy=name&sortOrder=ascending&groupBy=toolingTypes&licenses=&languages=&drafts=&toolingTypes=&environments=) has tools you can use to render the product type JSON schema and validate your product payload programmatically. 

Hope this helps! Let me know if you have any questions about the information above.

Weiling
Selling Partner Developer Services

**Comment by johndoeapi1-lab** [comment]
> I am integrating an external system to publish products on Amazon. I get type definitions for specific categories from the product type definitions API, which gives me the schema.
> 
> The problem is that I want to get all the required attributes and enums of those attributes ahead of time. But the schema is so complex. What is the solution to this specific problem? I appreciate your interest!

You found any better approach for it? Like how to get all the required attributes of a specific product-type?

**Reply by SylvieManx** [reply]
I spent a considerable amount of time researching how to handle listings across all categories on global websites, but I haven't made it public. I've shared a general description of the process here: https://zhuanlan.zhihu.com/p/705384610
